src/posts/2020-09-19-how-to-write-accessible-forms.md

```diff
@@ -29,69 +37,124 @@ tags:
+- The forms should be split into logical groups.
+- The forms should inform the person about the progress thay are making.
 
 
 ## Provide error messages if any
-If your site has forms, validation errors are inevitable. Always provide error messages when they occur. It is common practice to change the `<input>` color to red which is not ideal. Error messages helps people know exactly what they need to submit the form. 
 
-![A form with two fields indicating an error.](https://res.cloudinary.com/dgn6edv1k/image/upload/v1600197951/err_qiuycf.jpg)
+If your site has forms, validation errors are inevitable.
+When people encounter errors, clearly they need more help. Some people, like the visually impaired, may not even be aware that an error has occurred. It is common practice to change `<input>` color to red which is not ideal. This is because assistive devices will not catch the error, hence not informing the person of the error.
 
-## Avoid horizontal layout forms unless neccessary
-If possible, avoid making your form inputs next to each other. This allows people to follow the path to completion easily and focus on entering one piece of data at a time. 
 
-![A form with four input fields in a horizontal layout.](https://res.cloudinary.com/dgn6edv1k/image/upload/v1600201124/layout_glyubv.jpg)
+The following are ways to effectively and accessibly communicate form errors:
+- Let the person know there is an error.
+- Let the person access the error, and provide instructions on how to fix it.
+- Allow the person re-submit the form.
 
-If you must use a horizontal layout, make sure the tab order goes from left to right. This is done by adding a  `tabindex` attribute to `<input>`. The `tabindex` attribute should be in ascending order(1,2,3 and so on).
+
+In the code below, there is an empty [live region](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) which report errors to assistive devices. Additionally, each field is followed by a _hint_ which is hidden by default but shown if that field does not validate.
 
 ```html
+<form>
+  <fieldset>
+    <legend>Create a free account</legend>
+    <label for="username">Enter a username</label>
+    <input type="text" id="username">
+    <p class="hint hidden" id="usernameHint" >Your username cannot contain punctuation</p>
+
+    <label for="email">Enter your email address</label>
+    <input type="text" id="email">
+    <p class="hint hidden" id="emailHint">Please enter a valid email address</p>
+
+    <label for="password">Enter a password</label>
+    <input type="password" id="password">
+    <p class="hint hidden" id="passwordHint">Your password must be at least 6 characters</p>
+  </fieldset>
+
+  <div aria-live="assertive" id="message"></div>
+ 
+  <button type="submit">Submit</button>
+  
+</form>
+```
 
- <input tabindex="1" type="text" name="field1" />
- <input tabindex="2" type="text" name="field2" />
+```css
+.hidden{
+  text-indent: -5000px;

```

Comment:
in this instance, this is not a proper way to hide validation messaging.   

someone using a screen reader, navigating by the virtual cursor, would be able to come across these error messages as they traverse the DOM.  `display: none`, or simply NOT populating the containing element with text would be the best way to setup error messaging prior to the field being flagged as invalid.

Additionally, it would be worth noting that name and password specifically are introducing requirements that would otherwise be unknown to people trying to fill out this form.  This would fall into the area of failing 3.3.2 if these instructions were not provided _while_ the person was trying to fill out the form, rather than merely when they've already messed up.